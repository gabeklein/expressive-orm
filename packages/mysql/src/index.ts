export {
  default as Connection,
  default as MySQLConnection
} from './Connection';

export * from "@expressive/sql";