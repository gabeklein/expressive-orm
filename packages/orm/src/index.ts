export { default, Type } from './Type';
export type { Compat, Insert, Values } from './Type';
export * from './fields';
export * from './where';
export { one } from './one';
export { get } from './get';