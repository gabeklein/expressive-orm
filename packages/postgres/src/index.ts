export * from "@expressive/sql";
export { PGLiteConnection } from "./PGLiteConnection";
export { PGConnection } from "./PGConnection";