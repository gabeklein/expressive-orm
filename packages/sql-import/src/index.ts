export { Scanner } from './Scanner';
export { Parser } from './generate/Parser';