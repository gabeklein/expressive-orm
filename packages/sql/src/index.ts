export { Bool } from './field/Bool';
export { Connection } from './connection/Connection';
export { default as Util } from './utility';
export { Enum } from './field/Enum';
export { Field } from './Field';
export { Join } from './field/Join';
export { Json } from './field/Json';
export { Many } from './field/Many';
export { Num } from './field/Num';
export { One } from './field/One';
export { Query } from './query/Query';
export { Ref } from './field/Ref';
export { Scanner } from './parse/Scanner';
export { Schema } from './connection/Schema';
export { Str } from './field/Str';
export { Table } from './field/Table';
export { Time } from './field/Time';
export { Type } from './Type';