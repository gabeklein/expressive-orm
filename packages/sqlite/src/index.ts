import SQLiteConnection from './Connection';

export * from "@expressive/orm";

export {
  SQLiteConnection,
  SQLiteConnection as default
}