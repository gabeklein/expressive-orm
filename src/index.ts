import Bool from './columns/Bool';
import Int from './columns/Int';
import Many from './columns/Many';
import One from './columns/One';
import VarChar from './columns/VarChar';
import Entity from './entity';

class Author extends Entity {
  name = VarChar();
  age = Int();
  nickname = VarChar({ nullable: true });
  active = Bool();
  books = Many(Book);
}

class Book extends Entity {
  title = VarChar();
  author = One(Author); 
}

Author.register();
Book.register();

export async function getBook(){
  return Book.getOne({
    where(){
      this.title.is("1984");
      this.author.age.isMore(50);
    },
    select(){
      return this.title;
    }
  })
}

export async function getAuthor(){
  return Author.getOne({
    where(){
      this.name.isNot("gabe");
      this.nickname.is("gabe");
      this.age.isMore(3);
      this.active.is(true);
    },
    select(){
      return {
        name: this.nickname,
        age: this.age,
        status: this.active ? "active" : "inactive"
      } as const
    }
  });
}

console.clear()
getBook();