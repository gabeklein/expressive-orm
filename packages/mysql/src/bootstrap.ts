import { Field, Type, Util } from '@expressive/sql';

export function bootstrap(entities: Iterable<Type.EntityType>){
  const commands = [] as string[];
  
  for(const entity of entities)
    entity.ready();

  for(const entity of entities)
    commands.push(create(entity));

  for(const entity of entities){
    const statement = constraints(entity);

    if(statement)
      commands.push(statement);
  }
  
  return commands.join(";");
}

export function drop(table: Type.EntityType){
  return `DROP TABLE IF EXISTS ${table.name}`
}

export function create(table: Type.EntityType){
  const { table: name } = table;
  const statements = [] as string[];

  table.fields.forEach(field => {
    const sql = column(field);

    if(sql)
      statements.push(sql);
  });

  return `CREATE TABLE IF NOT EXISTS ${name} (${statements.join(",")})`
}

export function constraints(table: Type.EntityType){
  const statement = [] as string[];

  table.fields.forEach(field => {
    const { constraint } = field;

    if(constraint)
      statement.push(constraint);
  })

  if(statement.length)
    return `ALTER TABLE ${table.name} ${statement.join(", ")}`
}

function column(from: Field){
  if(from.datatype === undefined)
    return;

  let statement = `${Util.qualify(from.column)} ${from.datatype}`;

  if(!from.nullable)
    statement += " NOT NULL";

  if(from.default !== undefined)
    statement += ` DEFAULT ${Util.escapeString(from.default)}`;

  if(from.datatype == "INT" && from.increment)
    statement += " AUTO_INCREMENT";

  if(from.primary)
    statement += " PRIMARY KEY";

  return statement;
}