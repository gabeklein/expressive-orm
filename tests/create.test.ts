import { format } from 'sql-formatter';
import { Bool, Entity, Int, Many, One } from '../src';
import MySQLConnection from '../src/mysql/Connection';

it("will create tables", () => {
  class Author extends Entity {
    name = String();
    age = Int();
    nickname = String({ nullable: true });
    active = Bool();
    books = Many(Book);
    publisher = One(Publisher);
  }
  
  class Publisher extends Entity {
    name = String();
  }
  
  class Book extends Entity {
    title = String();
    author = One(Author);
    rating = Int();
  }

  const connection =
    new MySQLConnection({
      dry: true
    }, {
      Author,
      Book,
      Publisher
    });

  const sql = connection.createTables();

  expect(format(sql)).toMatchInlineSnapshot(`
    CREATE TABLE
      IF NOT EXISTS Author (
        \`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
        \`age\` INT NOT NULL,
        \`active\` TINYINT NOT NULL,
        \`publisherId\` INT NOT NULL
      );

    CREATE TABLE
      IF NOT EXISTS Book (
        \`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
        \`authorId\` INT NOT NULL,
        \`rating\` INT NOT NULL
      );

    CREATE TABLE
      IF NOT EXISTS Publisher (\`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY);

    ALTER TABLE
      Author
    ADD
      CONSTRAINT \`FK_PublisherAuthor\` FOREIGN KEY (publisherId) REFERENCES Publisher(id);

    ALTER TABLE
      Book
    ADD
      CONSTRAINT \`FK_AuthorBook\` FOREIGN KEY (authorId) REFERENCES Author(id)
`);
})