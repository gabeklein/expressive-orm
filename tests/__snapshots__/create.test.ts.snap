// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will create tables 1`] = `
	CREATE TABLE
	  IF NOT EXISTS Author (
	    \`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
	    \`age\` INT NOT NULL,
	    \`active\` TINYINT NOT NULL,
	    \`publisherId\` INT NOT NULL
	  );
	
	CREATE TABLE
	  IF NOT EXISTS Book (
	    \`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
	    \`authorId\` INT NOT NULL,
	    \`rating\` INT NOT NULL
	  );
	
	CREATE TABLE
	  IF NOT EXISTS Publisher (\`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY);
	
	ALTER TABLE
	  Author ADD CONSTRAINT \`FK_PublisherAuthor\` FOREIGN KEY (publisherId) REFERENCES Publisher (id);
	
	ALTER TABLE
	  Book ADD CONSTRAINT \`FK_AuthorBook\` FOREIGN KEY (authorId) REFERENCES Author (id)
`;
